launch_pytest support for --reruns or --retest-until-pass

### Operating System:

Ubuntu 24

### ROS version or commit hash:

jazzy

### RMW implementation (if applicable):

_No response_

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

_No response_

### 'ros2 doctor --report' output

_No response_

### Steps to reproduce issue

When creating a `launch_pytest` test and running it with `--retry` from `pytest-rerunfailures` there appears to be an issue with how the even_loop is handled

1. Clone this repo (rolling)
2. `colcon build` and source
3. In the `launch_pytest/test/launch_pytest/examples/function_scope.py` file, Add a test case that failes some of the time:
```
@pytest.mark.launch(fixture=launch_description)
async def test_case_1(order):
    order.append('test_case_1')
    assert True
    import random
    assert random.choice([True, False, False, False, False])
```
4. `pip install pytest-rerunfailures` (this is what colcon uses internally)
5. Run the test with retry: `python3 -m pytest --reruns 10 --verbose ./launch_pytest/test/launch_pytest/examples/function_scope.py`

Tests always fail due to `RuntimeError: Event loop is closed`
```
platform linux -- Python 3.10.12, pytest-7.1.3, pluggy-1.0.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/david/launch/launch_pytest, configfile: pytest.ini
plugins: launch_pytest-3.8.0, launch_testing-3.8.0, ament-copyright-0.14.3, ament-xmllint-0.14.3, ament-lint-0.14.3, ament-pep257-0.14.3, launch-testing-ros-0.24.1, ament-flake8-0.14.3, rerunfailures-10.2, colcon-core-0.15.1, repeat-0.9.1, anyio-4.0.0, cov-3.0.0, typeguard-4.3.0
collected 6 items                                                                                                                                                                                        

launch_pytest/test/launch_pytest/examples/function_scope.py::test_after_shutdown PASSED                                                                                                            [ 16%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_1 RERUN                                                                                                                     [ 33%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_1 RERUN                                                                                                                     [ 33%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_1 RERUN                                                                                                                     [ 33%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_1 RERUN                                                                                                                     [ 33%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_1 RERUN                                                                                                                     [ 33%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_1 RERUN                                                                                                                     [ 33%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_1 RERUN                                                                                                                     [ 33%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_1 RERUN                                                                                                                     [ 33%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_1 RERUN                                                                                                                     [ 33%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_1 RERUN                                                                                                                     [ 33%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_1 FAILED                                                                                                                    [ 33%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_2 PASSED                                                                                                                    [ 50%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_3 PASSED                                                                                                                    [ 66%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_4 PASSED                                                                                                                    [ 83%]
launch_pytest/test/launch_pytest/examples/function_scope.py::test_order PASSED                                                                                                                     [100%]

================================================================================================ FAILURES ================================================================================================
______________________________________________________________________________________________ test_case_1 _______________________________________________________________________________________________

kwargs = {'order': ['test_after_shutdown', 'test_case_1']}, future = <Future finished exception=RuntimeError('Event loop is closed')>

    @functools.wraps(func)
    def inner(**kwargs):
        if before_test is not None:
            before_test()
        future = event_loop.run_in_executor(None, functools.partial(func, **kwargs))
>       run_until_complete(event_loop, future)

install/launch_pytest/lib/python3.10/site-packages/launch_pytest/plugin.py:396: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
install/launch_pytest/lib/python3.10/site-packages/launch_pytest/plugin.py:369: in run_until_complete
    loop.run_until_complete(future_like)
/usr/lib/python3.10/asyncio/base_events.py:649: in run_until_complete
    return future.result()
/usr/lib/python3.10/concurrent/futures/thread.py:58: in run
    result = self.fn(*self.args, **self.kwargs)
install/launch_pytest/lib/python3.10/site-packages/launch_pytest/plugin.py:395: in inner
    future = event_loop.run_in_executor(None, functools.partial(func, **kwargs))
/usr/lib/python3.10/asyncio/base_events.py:808: in run_in_executor
    self._check_closed()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_UnixSelectorEventLoop running=False closed=True debug=False>

    def _check_closed(self):
        if self._closed:
>           raise RuntimeError('Event loop is closed')
E           RuntimeError: Event loop is closed

/usr/lib/python3.10/asyncio/base_events.py:515: RuntimeError
----------------------------------------------------------------------------------------- Captured stdout setup ------------------------------------------------------------------------------------------
[INFO] [launch]: All log files can be found below /home/david/.ros/log/2025-04-09-15-51-51-288218-david-297144
[INFO] [launch]: Default logging verbosity is set to INFO
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-2]: process started with pid [297147]
[INFO] [python3-2]: sending signal 'SIGINT' to process[python3-2]
[INFO] [python3-2]: process has finished cleanly [pid 297147]
----------------------------------------------------------------------------------------- Captured stdout setup ------------------------------------------------------------------------------------------
[INFO] [launch]: All log files can be found below /home/david/.ros/log/2025-04-09-15-51-51-288218-david-297144
[INFO] [launch]: Default logging verbosity is set to INFO
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-3]: process started with pid [297149]
[INFO] [python3-3]: sending signal 'SIGINT' to process[python3-3]
[INFO] [python3-3]: process has finished cleanly [pid 297149]
----------------------------------------------------------------------------------------- Captured stdout setup ------------------------------------------------------------------------------------------
[INFO] [launch]: All log files can be found below /home/david/.ros/log/2025-04-09-15-51-51-288218-david-297144
[INFO] [launch]: Default logging verbosity is set to INFO
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-4]: process started with pid [297152]
[INFO] [python3-4]: sending signal 'SIGINT' to process[python3-4]
[INFO] [python3-4]: process has finished cleanly [pid 297152]
----------------------------------------------------------------------------------------- Captured stdout setup ------------------------------------------------------------------------------------------
[INFO] [launch]: All log files can be found below /home/david/.ros/log/2025-04-09-15-51-51-288218-david-297144
[INFO] [launch]: Default logging verbosity is set to INFO
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-5]: process started with pid [297155]
[INFO] [python3-5]: sending signal 'SIGINT' to process[python3-5]
[INFO] [python3-5]: process has finished cleanly [pid 297155]
----------------------------------------------------------------------------------------- Captured stdout setup ------------------------------------------------------------------------------------------
[INFO] [launch]: All log files can be found below /home/david/.ros/log/2025-04-09-15-51-51-288218-david-297144
[INFO] [launch]: Default logging verbosity is set to INFO
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-6]: process started with pid [297158]
[INFO] [python3-6]: sending signal 'SIGINT' to process[python3-6]
[INFO] [python3-6]: process has finished cleanly [pid 297158]
----------------------------------------------------------------------------------------- Captured stdout setup ------------------------------------------------------------------------------------------
[INFO] [launch]: All log files can be found below /home/david/.ros/log/2025-04-09-15-51-51-288218-david-297144
[INFO] [launch]: Default logging verbosity is set to INFO
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-7]: process started with pid [297161]
[INFO] [python3-7]: sending signal 'SIGINT' to process[python3-7]
[INFO] [python3-7]: process has finished cleanly [pid 297161]
----------------------------------------------------------------------------------------- Captured stdout setup ------------------------------------------------------------------------------------------
[INFO] [launch]: All log files can be found below /home/david/.ros/log/2025-04-09-15-51-51-288218-david-297144
[INFO] [launch]: Default logging verbosity is set to INFO
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-8]: process started with pid [297164]
[INFO] [python3-8]: sending signal 'SIGINT' to process[python3-8]
[INFO] [python3-8]: process has finished cleanly [pid 297164]
----------------------------------------------------------------------------------------- Captured stdout setup ------------------------------------------------------------------------------------------
[INFO] [launch]: All log files can be found below /home/david/.ros/log/2025-04-09-15-51-51-288218-david-297144
[INFO] [launch]: Default logging verbosity is set to INFO
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-9]: process started with pid [297167]
[INFO] [python3-9]: sending signal 'SIGINT' to process[python3-9]
[INFO] [python3-9]: process has finished cleanly [pid 297167]
----------------------------------------------------------------------------------------- Captured stdout setup ------------------------------------------------------------------------------------------
[INFO] [launch]: All log files can be found below /home/david/.ros/log/2025-04-09-15-51-51-288218-david-297144
[INFO] [launch]: Default logging verbosity is set to INFO
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-10]: process started with pid [297170]
[INFO] [python3-10]: sending signal 'SIGINT' to process[python3-10]
[INFO] [python3-10]: process has finished cleanly [pid 297170]
----------------------------------------------------------------------------------------- Captured stdout setup ------------------------------------------------------------------------------------------
[INFO] [launch]: All log files can be found below /home/david/.ros/log/2025-04-09-15-51-51-288218-david-297144
[INFO] [launch]: Default logging verbosity is set to INFO
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-11]: process started with pid [297173]
[INFO] [python3-11]: sending signal 'SIGINT' to process[python3-11]
[INFO] [python3-11]: process has finished cleanly [pid 297173]
----------------------------------------------------------------------------------------- Captured stdout setup ------------------------------------------------------------------------------------------
[INFO] [launch]: All log files can be found below /home/david/.ros/log/2025-04-09-15-51-51-288218-david-297144
[INFO] [launch]: Default logging verbosity is set to INFO
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-2]: process started with pid [297147]
[INFO] [python3-2]: sending signal 'SIGINT' to process[python3-2]
[INFO] [python3-2]: process has finished cleanly [pid 297147]
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-3]: process started with pid [297149]
[INFO] [python3-3]: sending signal 'SIGINT' to process[python3-3]
[INFO] [python3-3]: process has finished cleanly [pid 297149]
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-4]: process started with pid [297152]
[INFO] [python3-4]: sending signal 'SIGINT' to process[python3-4]
[INFO] [python3-4]: process has finished cleanly [pid 297152]
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-5]: process started with pid [297155]
[INFO] [python3-5]: sending signal 'SIGINT' to process[python3-5]
[INFO] [python3-5]: process has finished cleanly [pid 297155]
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-6]: process started with pid [297158]
[INFO] [python3-6]: sending signal 'SIGINT' to process[python3-6]
[INFO] [python3-6]: process has finished cleanly [pid 297158]
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-7]: process started with pid [297161]
[INFO] [python3-7]: sending signal 'SIGINT' to process[python3-7]
[INFO] [python3-7]: process has finished cleanly [pid 297161]
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-8]: process started with pid [297164]
[INFO] [python3-8]: sending signal 'SIGINT' to process[python3-8]
[INFO] [python3-8]: process has finished cleanly [pid 297164]
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-9]: process started with pid [297167]
[INFO] [python3-9]: sending signal 'SIGINT' to process[python3-9]
[INFO] [python3-9]: process has finished cleanly [pid 297167]
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-10]: process started with pid [297170]
[INFO] [python3-10]: sending signal 'SIGINT' to process[python3-10]
[INFO] [python3-10]: process has finished cleanly [pid 297170]
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-11]: process started with pid [297173]
[INFO] [python3-11]: sending signal 'SIGINT' to process[python3-11]
[INFO] [python3-11]: process has finished cleanly [pid 297173]
---------------------------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------------------------
[INFO] [python3-12]: process started with pid [297176]
[INFO] [python3-12]: sending signal 'SIGINT' to process[python3-12]
[INFO] [python3-12]: process has finished cleanly [pid 297176]
======================================================================================== short test summary info =========================================================================================
FAILED launch_pytest/test/launch_pytest/examples/function_scope.py::test_case_1 - RuntimeError: Event loop is closed
```



### Expected behavior


python3 -m pytest ./packages/geopose_history/ --verbose --reruns 2 --reruns-delay 1

### Actual behavior

On the balance of probabilities, this test should pass after ~10 retries

### Additional information

_No response_

--- a/launch_pytest/launch_pytest/plugin.py
+++ b/launch_pytest/launch_pytest/plugin.py
@@ -316,7 +316,12 @@ def pytest_pyfunc_call(pyfuncitem):
         yield
         return
 
-    func = pyfuncitem.obj
+    # Store the original unwrapped function to avoid re-wrapping on reruns.
+    # This prevents issues with pytest plugins like pytest-flaky or pytest-rerunfailures,
+    # which reuse the same pyfuncitem across multiple test runs.
+    func = getattr(pyfuncitem, "_launch_pytest_original_obj", pyfuncitem.obj)
+    pyfuncitem._launch_pytest_original_obj = func
+    
     if has_shutdown_kwarg(pyfuncitem) and need_shutdown_test_item(func):
         error_msg = (
             'generator or async generator based launch test items cannot be marked with'

--- a/launch_pytest/launch_pytest/plugin.py
+++ b/launch_pytest/launch_pytest/plugin.py
@@ -319,9 +319,9 @@ def pytest_pyfunc_call(pyfuncitem):
     # Store the original unwrapped function to avoid re-wrapping on reruns.
     # This prevents issues with pytest plugins like pytest-flaky or pytest-rerunfailures,
     # which reuse the same pyfuncitem across multiple test runs.
-    func = getattr(pyfuncitem, "_launch_pytest_original_obj", pyfuncitem.obj)
+    func = getattr(pyfuncitem, '_launch_pytest_original_obj', pyfuncitem.obj)
     pyfuncitem._launch_pytest_original_obj = func
-    
+
     if has_shutdown_kwarg(pyfuncitem) and need_shutdown_test_item(func):
         error_msg = (
             'generator or async generator based launch test items cannot be marked with'

